src/audio/template_comp/template_comp.c

```diff
@@ -0,0 +1,309 @@
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		return -ENOMEM;
+
+	md->private = cd;
+	return 0;
+}
+
+__cold static int template_comp_free(struct processing_module *mod)
+{
+	struct template_comp_comp_data *cd = module_get_private_data(mod);
+
+	assert_can_be_cold();
+
+	comp_info(mod->dev, "template_comp_free()");
+	rfree(cd);
+	return 0;
+}
+
+#if CONFIG_FORMAT_S16LE
+static void template_comp_s16(const struct processing_module *mod,
+			      const struct audio_stream *source,
+			      struct audio_stream *sink,
+			      uint32_t frames)
+{
+	struct template_comp_comp_data *cd = module_get_private_data(mod);
+	int16_t *x = audio_stream_get_rptr(source);
+	int16_t *y = audio_stream_get_wptr(sink);
+	int nbuf;
+	int nfrm;
+	int ch;
+	int i;
+
+	while (frames) {
+		nbuf = audio_stream_frames_without_wrap(source, x);
+		nfrm = MIN(frames, nbuf);
+		nbuf = audio_stream_frames_without_wrap(sink, y);
+		nfrm = MIN(nfrm, nbuf);
+		for (i = 0; i < nfrm; i++) {
+			for (ch = 0; ch < cd->channels; ch++) {
+				*y = *(x + cd->channels_order[ch]);
+				y++;
+			}
+			x += cd->channels;
+		}
+		frames -= nfrm;
+		x = audio_stream_wrap(source, x);
+		y = audio_stream_wrap(sink, y);
+	}
+}
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S32LE
+/* Same function works for s24 and s32 since
+ * the samples values are not modified.
+ */
+static void template_comp_s32(const struct processing_module *mod,
+			      const struct audio_stream *source,
+			      struct audio_stream *sink,
+			      uint32_t frames)
+{
+	struct template_comp_comp_data *cd = module_get_private_data(mod);
+	int32_t *x = audio_stream_get_rptr(source);
+	int32_t *y = audio_stream_get_wptr(sink);
+	int nbuf;
+	int nfrm;
+	int ch;
+	int i;
+
+	while (frames) {
+		nbuf = audio_stream_frames_without_wrap(source, x);
+		nfrm = MIN(frames, nbuf);
+		nbuf = audio_stream_frames_without_wrap(sink, y);
+		nfrm = MIN(nfrm, nbuf);
+		for (i = 0; i < nfrm; i++) {
+			for (ch = 0; ch < cd->channels; ch++) {
+				*y = *(x + cd->channels_order[ch]);
+				y++;
+			}
+			x += cd->channels;
+		}
+		frames -= nfrm;
+		x = audio_stream_wrap(source, x);
+		y = audio_stream_wrap(sink, y);
+	}
+}
+#endif /* CONFIG_FORMAT_S32LE */
+
+const struct template_comp_proc_fnmap template_comp_proc_fnmap[] = {
+#if CONFIG_FORMAT_S16LE
+	{ SOF_IPC_FRAME_S16_LE, template_comp_s16 },
+#endif
+#if CONFIG_FORMAT_S24LE
+	{ SOF_IPC_FRAME_S24_4LE, template_comp_s32 },
+#endif
+#if CONFIG_FORMAT_S32LE
+	{ SOF_IPC_FRAME_S32_LE, template_comp_s32 },
+#endif
+};
+
+static inline template_comp_func template_comp_find_proc_func(enum sof_ipc_frame src_fmt)
+{
+	int i;
+
+	/* Find suitable processing function from map */
+	for (i = 0; i < ARRAY_SIZE(template_comp_proc_fnmap); i++)
+		if (src_fmt == template_comp_proc_fnmap[i].frame_fmt)
+			return template_comp_proc_fnmap[i].template_comp_proc_func;
+
+	return NULL;
+}
+
+static int template_comp_process(struct processing_module *mod,
```

Comment:
I strongly suggest using sink/source in the template for new modules

Reply:
I've converted it now to sink/source. But as not experienced with it, not sure if the best way for everything. Some things look now a bit more complex than before. Can you please check again this!

Reply:
@marcinszkudlinski has your request been addressed correctly? Should your request for change be dropped?